contracts/Presale.sol

```diff
@@ -0,0 +1,227 @@
+        return $rounds;
+    }
+
+    function totalRounds() external view override returns (uint256) {
+        return $rounds.length;
+    }
+
+    function raisedUSD(uint256 roundIndex) external view returns (uint256) {
+        return $raisedUSD[roundIndex];
+    }
+
+    function totalRaisedUSD() external view override returns (uint256) {
+        return $totalRaisedUSD;
+    }
+
+    function userTokensAllocated(address account) external view override returns (uint256) {
+        return $userTokensAllocated[account];
+    }
+
+    function ethPrice() public view returns (uint256) {
+        (, int256 price,,,) = ORACLE.latestRoundData();
+        return uint256(uint256(price) * ETH_TO_WEI_PRECISION);
+    }
+
+    function ethToUsd(uint256 amount) public view returns (uint256) {
+        uint256 _ethPrice = ethPrice();
+        uint256 _ethAmountInUsd = (_ethPrice * amount) / PRECISION;
+        return _ethAmountInUsd;
+    }
+
+    function usdToTokens(uint256 roundIndex, uint256 amount) public view returns (uint256) {
+        RoundConfig memory _round = $rounds[roundIndex];
+        return amount * _round.tokenPrice;
+    }
+
+    function setConfig(PresaleConfig calldata newConfig) external override onlyOwner {
+        require(newConfig.startDate < newConfig.endDate, "INVALID_DATES");
+        require(newConfig.withdrawTo != address(0), "INVALID_WITHDRAW_TO");
+
+        emit ConfigUpdated($config, newConfig, _msgSender());
+
+        $config = newConfig;
+    }
+
+    function setRounds(RoundConfig[] calldata newRounds) external override onlyOwner {
+        emit RoundsUpdated($rounds, newRounds, _msgSender());
+
+        uint256 _totalRaisedUSD = $totalRaisedUSD;
+
+        uint256 totalCostUSD;
+        uint256 expectedCurrentRoundIndex;
+
+        for (uint256 i; i < newRounds.length; ++i) {
+            $rounds.push(newRounds[i]);
+
+            uint256 roundCostUSD = newRounds[i].tokensAllocated * newRounds[i].tokenPrice;
+            totalCostUSD += roundCostUSD;
+            if (_totalRaisedUSD > totalCostUSD) {
+                expectedCurrentRoundIndex++;
+            }
+        }
+
+        $currentRoundIndex = expectedCurrentRoundIndex;
+    }
+
+    function purchase() public payable override whenNotPaused {
+        uint256 amountUSD = ethToUsd(msg.value);
+        _sync($currentRoundIndex, address(0), amountUSD, _msgSender());
+    }
+
+    function purchase(address account) public payable override whenNotPaused {
+        uint256 amountUSD = ethToUsd(msg.value);
+        _sync($currentRoundIndex, address(0), amountUSD, account);
+    }
+
+    function purchaseUSDC(uint256 amount) external override whenNotPaused {
+        address sender = _msgSender();
+
+        IERC20(USDC).transferFrom(sender, address(this), amount);
+
+        uint256 amountScaled = amount * USDC_TO_WEI_PRECISION;
+        _sync($currentRoundIndex, USDC, amountScaled, sender);
```

Comment:
after sync for each purchase, we can transfer the remaining assets that we hold to the withdrawal address. As refund should have happened by now.

Reply:
added return value to _sync of total purchase amount in asset, and transferred this amount to `withdrawTo`